File: 01-programming-languages/PHP/02-oop/03-traits/traits.php
<?php
// PHP Traits

trait Logger {
    public function log(string $msg): void {
        echo "LOG: $msg\n";
    }
}

trait Timestamp {
    public function timestamp(): string {
        return date('Y-m-d H:i:s');
    }
}

class Service {
    use Logger, Timestamp;

    public function doWork(): void {
        $this->log("starting at " . $this->timestamp());
        echo "working\n";
        $this->log("done at " . $this->timestamp());
    }
}

$service = new Service();
$service->doWork();

// Trait conflict resolution
trait A {
    public function hello(): void {
        echo "A\n";
    }
}

trait B {
    public function hello(): void {
        echo "B\n";
    }
}

class C {
    use A, B {
        A::hello insteadof B;
        B::hello as world;
    }
}

$c = new C();
$c->hello();
$c->world();

File: 01-programming-languages/Scala/02-advanced/03-traits/traits.scala
// Traits in Scala

trait Drawable {
  def draw(): Unit
}

trait Resizable {
  def resize(factor: Double): Unit
}

class Circle(val radius: Double) extends Drawable with Resizable {
  def draw(): Unit = println(s"Drawing circle with radius $radius")
  def resize(factor: Double): Unit = println(s"Resizing circle by $factor")
}

class Square(val side: Double) extends Drawable {
  def draw(): Unit = println(s"Drawing square with side $side")
}

trait Logger {
  def log(msg: String): Unit = println(s"LOG: $msg")
}

class Service extends Logger {
  def doWork(): Unit = {
    log("starting work")
    println("doing work")
    log("work done")
  }
}

object Traits {
  def main(args: Array[String]): Unit = {
    val circle = Circle(5.0)
    circle.draw()
    circle.resize(2.0)

    val square = Square(3.0)
    square.draw()

    val service = new Service
    service.doWork()
  }
}
